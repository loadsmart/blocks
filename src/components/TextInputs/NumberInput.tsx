import React, { PureComponent } from 'react'
import {
  NativeSyntheticEvent,
  Platform,
  StyleSheet,
  TextInput,
  TextInputKeyPressEventData,
  TextInputProps,
} from 'react-native'
import { Colors } from '../../res'
import { ThemeContext } from '../Contexts'

interface Props extends TextInputProps {
  onChangeText: (text: string) => void
  digit?: string
  focus?: boolean
}

interface State {
  focused: boolean
}

const digitRegex = '^\\d$'

export default class NumberInput extends PureComponent<Props, State> {
  private textInput: TextInput | null

  constructor(props: Props, state: State) {
    super(props, state)
    this.textInput = null
    this.state = {
      focused: false,
    }
  }

  public render() {
    return (
      <ThemeContext.Consumer>
        {theme => {
          const borderStyle = this.state.focused
            ? { borderColor: theme.primaryColor }
            : { borderColor: Colors.Ghost }
          return (
            <TextInput
              ref={component => (this.textInput = component)}
              keyboardType={Platform.OS === 'ios' ? 'number-pad' : 'default'}
              maxLength={1}
              selectionColor={Colors.Transparent}
              style={[styles.container, borderStyle]}
              onChangeText={this.onChangeText}
              onBlur={this.onBlur}
              onFocus={this.onFocus}
              blurOnSubmit={false}
              value={this.props.digit}
              onKeyPress={this.onKeyPress}
              caretHidden={true}
            />
          )
        }}
      </ThemeContext.Consumer>
    )
  }

  public onChangeText = (text: string) => {
    if (text.match(digitRegex)) {
      this.props.onChangeText(text)
    }
  }

  public blur = () => {
    this.textInput!.blur()
  }

  public focus = () => {
    this.textInput!.focus()
  }

  public onBlur = () => {
    this.setState({ focused: false })
  }

  public onFocus = () => {
    this.setState({ focused: true })
  }

  public onKeyPress = (event: NativeSyntheticEvent<TextInputKeyPressEventData>) => {
    const key = event.nativeEvent.key
    if (key.match(digitRegex)) {
      this.props.onChangeText(key)
    } else if (key.match('Backspace')) {
      this.props.onChangeText('')
    }
  }
}

const styles = StyleSheet.create({
  container: {
    width: 45,
    height: 63,
    borderWidth: 2,
    backgroundColor: Colors.White,
    fontSize: 20,
    textAlign: 'center',
    marginHorizontal: 4,
  },
})
