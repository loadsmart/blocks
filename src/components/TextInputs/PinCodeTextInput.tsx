import React, { Component } from 'react'
import { StyleSheet, TextInputProps, View } from 'react-native'
import NumberInput from './NumberInput'

interface Props extends TextInputProps {
  onChangeInput: (input: string) => void
  authCode?: string
}

interface State {
  digits: string[]
  focused: boolean
}

const numberOfDigits = 6

export default class PinCodeTextInput extends Component<Props, State> {
  private inputRefs: NumberInput[]

  constructor(props: Props, state: State) {
    super(props, state)
    this.inputRefs = []
    this.state = {
      digits: [...Array(numberOfDigits)].map(_ => ''),
      focused: false,
    }
  }

  public static getDerivedStateFromProps(nextProps: Props, prevState: State) {
    if (nextProps.authCode === prevState.digits.join('') || nextProps.authCode === undefined) {
      return prevState
    }

    return {
      digits: nextProps.authCode!.split(''),
    }
  }

  public render() {
    return (
      <View style={styles.pinNumbersContainer}>
        {[...Array(numberOfDigits).keys()].map(i => (
          <NumberInput
            {...this.props}
            key={i}
            ref={component => (this.inputRefs[i] = component!)}
            onChangeText={text => {
              this.onChangeDigit(i, text)
            }}
            digit={this.state.digits[i]}
          />
        ))}
      </View>
    )
  }

  public onChangeDigit = (index: number, text: string) => {
    this.state.digits[index] = text
    this.props.onChangeInput(this.state.digits.join(''))
    this.updatePinCodeInputFocus(index, this.inputRefs, text)
  }

  public updatePinCodeInputFocus = (index: number, inputRefs: NumberInput[], text: string) => {
    const isDeleting = text.length === 0
    if (isDeleting) {
      inputRefs[index].blur()
      if (index > 0) {
        inputRefs[index - 1].focus()
      }
    } else {
      inputRefs[index].blur()
      if (index !== inputRefs.length - 1) {
        inputRefs[index + 1].focus()
      }
    }
  }
}

const styles = StyleSheet.create({
  pinNumbersContainer: {
    flexDirection: 'row',
  },
})
