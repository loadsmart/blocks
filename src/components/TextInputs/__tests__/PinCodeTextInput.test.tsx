import React from 'react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import NumberInput from '../NumberInput'
import PinCodeTextInput from '../PinCodeTextInput'

describe('PinCodeTextInput', () => {
  let testRenderer: ReactTestRenderer

  const renderPinCodeComponent = (authCode?: string) => {
    testRenderer = renderer.create(
      <PinCodeTextInput authCode={authCode} onChangeInput={jest.fn()} />
    )
  }

  describe('when auth code prop is defined', () => {
    it('renders all number inputs with a digit', () => {
      renderPinCodeComponent('123456')
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when auth code prop is undefined', () => {
    it('renders all number inputs with a empty value', () => {
      renderPinCodeComponent()
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when auth code changes', () => {
    it('renders a number input with a valid digit', () => {
      renderPinCodeComponent('123456')
      testRenderer.root.instance.onChangeDigit(0, '1')
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when adding or updating a digit', () => {
    it('renders the number input with the a new digit', () => {
      renderPinCodeComponent('123456')
      const numberInputs = testRenderer.root.findAllByType(NumberInput)
      numberInputs.forEach(input => input.props.onChangeText('1'))
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('when deleting a digit', () => {
    it('renders the number input with empty value', () => {
      renderPinCodeComponent('123456')
      const numberInputs = testRenderer.root.findAllByType(NumberInput)
      numberInputs.forEach(input => input.props.onChangeText(''))
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })
})
