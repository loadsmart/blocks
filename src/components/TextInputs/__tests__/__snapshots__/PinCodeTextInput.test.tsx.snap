// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinCodeTextInput render renders all number inputs with a digit if auth code prop is defined 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="1"
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="2"
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="3"
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="4"
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="5"
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="6"
  />
</View>
`;

exports[`PinCodeTextInput render renders all number inputs with a empty value if auth code prop is undefined 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={false}
    caretHidden={true}
    keyboardType="number-pad"
    maxLength={1}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    rejectResponderTermination={true}
    selectionColor="#00000000"
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderWidth": 2,
          "fontSize": 20,
          "height": 63,
          "marginHorizontal": 4,
          "textAlign": "center",
          "width": 45,
        },
        Object {
          "borderColor": "#c4ccd6",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
